### DISCUSSION TITLE
Equivalence checking between the C/C++ and GOTO program

### DISCUSSION THREAD

**Discussion by hassan-truscova** [esbmc/esbmc#1897] [General]
Hi,

i was wondering if you perform equivalence checking between the C/C++ abstraction and the GOTO program?
I could not find it in the papers or anywhere else. How is it ensured that the generated GOTO program is equivalent to the C/C++ program?

Best wishes,

**Comment by lucasccordeiro** [comment]
Dear @hassan-truscova,

Our research group has never tried to check the formal equivalence between the C/C++ abstract and the GOTO program.

Can I ask the reason by which you are interested in this equivalence?

ESBMC can export the GOTO program as a C program, which can be used to check the equivalence between the original C/C++ program and the converted GOTO program.

For example, consider this C code fragment:

````C
int func_orig(_Bool a, _Bool b){

  if (a)
    return f();
  else if (b)
    return g();
  else return h();

  return 0;
}
````

If you call ESBMC with the option `--goto2c`, you'll get this converted GOTO program in C:

````C
int  func_goto(_Bool a, _Bool b)
{
if(!a) goto __ESBMC_goto_label_1; //  // GOTO
return f(); //  // FUNCTION_CALL
goto __ESBMC_goto_label_3; //  // GOTO
__ESBMC_goto_label_1:; // Target
if(!b) goto __ESBMC_goto_label_2; //  // GOTO
return g(); //  // FUNCTION_CALL
goto __ESBMC_goto_label_3; //  // GOTO
__ESBMC_goto_label_2:; // Target
return h(); //  // FUNCTION_CALL
__ESBMC_goto_label_3:; // Target
return 0; //  // RETURN
// main; //  // END_FUNCTION
}
````

You can check the equivalence as follows:

````C
int main(){
  _Bool a, b;
  assert(func_goto(a,b) == func_orig(a,b));
  return 0;
}
````

**Reply by hassan-truscova** [reply]
Dear @lucasccordeiro,

I was asking because of two reasons, 1) i couldn't find this information in any research paper or documentation, and 2) i wanted to extend it towards Solidity. I have already tried the approach suggested by you and somehow the model checker always gives me equivalence. Maybe, i am not using the correct switches, hence, you can help me find the correct answer maybe :)

I have the following example function which i took from the [documentation](https://ssvlab.github.io/esbmc/documentation.html#how-to-run)

```c++
int foo_original(unsigned int N, double x) {
  
  if(x <= 0 || isnan(x))
    return 0;
  unsigned int i = 0;
  while(i < N) {
    x = (2*x);
    assert(x>0);
    ++i;
  }
  assert(x>0);
  return 0;
}
```

I convert it into GOTO and get the following

```c++
int  foo(unsigned int N, double x)
{
if(!(x <= 0.000000 || isnan(x))) goto __ESBMC_goto_label_1; //  // GOTO
return 0; //  // RETURN
__ESBMC_goto_label_1:; // Target
unsigned int i; //  // DECL
i=0; //  // ASSIGN
__ESBMC_goto_label_2:; // Target
if(!(i < N)) goto __ESBMC_goto_label_3; //  // GOTO
x=(2.000000)*(x); //  // ASSIGN
assert(x > 0.000000); //  // ASSERT
i=i + 1; //  // ASSIGN
goto __ESBMC_goto_label_2; //  // GOTO
__ESBMC_goto_label_3:; // Target
assert(x > 0.000000); //  // ASSERT
return 0; //  // RETURN
// foo; //  // END_FUNCTION
}
```


Afterwards, i do the equivalence checking as follows:


```c++
int main () {

  unsigned int N;// = nondet_uint();
  double x;// = nondet_double();
  assert (foo (N,x) == foo_original (N,x));
  
  return 0;

}
```

With this flow, everything seems to work fine. :)
Now, i introduce a mutation in the function "foo_original":

```c++
int foo_original(unsigned int N, double x) {
  
  if(x <= 6 || isnan(x)) // <<<<<< original (x <= 0 || isnan(x))
    return 0;
  unsigned int i = 0;
  while(i < N) {
    x = (2*x);
    assert(x>0);
    ++i;
  }
  assert(x>0);
  return 0;
}
```

I do not generate a new GOTO, rather just do the equivalence check as before. The assertion should fail however, it is still successful. I am having some difficulty understanding this behaviour. There could be some misunderstanding on my part or maybe i need to use some extra switches. Current i use the following 

`esbmc file.c --floatbv --k-induction`

One additional thing, for some mutations in the function "foo_original", i do see a counter example.

Could you please guide me what i might be doing wrong in this scenario?

**Comment by lucasccordeiro** [comment]
Dear @hassan-truscova,

Many thanks for your response.

I think the main issue is that both functions just return 0. If you differentiate the return value, you'll get what you expect. For example, we can return -1 for the first if-statement to indicate that something went wrong and keep the return 0 at the end of the function to indicate that we could compute the result. 

````C
int foo_original(unsigned int N, double x) {

  if(x <= 0 || isnan(x))
    return -1;
  unsigned int i = 0;
  while(i < N) {
    x = (2*x);
    assert(x>0);
    ++i;
  }
  assert(x>0);
  return 0;
}

int  foo(unsigned int N, double x)
{
if(!(x <= 0.000000 || isnan(x))) goto __ESBMC_goto_label_1; //  // GOTO
return -1; //  // RETURN
__ESBMC_goto_label_1:; // Target
unsigned int i; //  // DECL
i=0; //  // ASSIGN
__ESBMC_goto_label_2:; // Target
if(!(i < N)) goto __ESBMC_goto_label_3; //  // GOTO
x=(2.000000)*(x); //  // ASSIGN
assert(x > 0.000000); //  // ASSERT
i=i + 1; //  // ASSIGN
goto __ESBMC_goto_label_2; //  // GOTO
__ESBMC_goto_label_3:; // Target
assert(x > 0.000000); //  // ASSERT
return 0; //  // RETURN
// foo; //  // END_FUNCTION
}

int main () {

  unsigned int N;// = nondet_uint();
  double x;// = nondet_double();
  assert (foo (N,x) == foo_original (N,x));

  return 0;

}
````

$esbmc equivalence2.c --floatbv --k-induction

````
...
Checking inductive step, k = 2
Starting Bounded Model Checking
Unwinding loop 3 iteration 1   file equivalence2.c line 27 column 1 function foo
Not unwinding loop 3 iteration 2   file equivalence2.c line 27 column 1 function foo
Unwinding loop 2 iteration 1   file equivalence2.c line 6 column 3 function foo_original
Not unwinding loop 2 iteration 2   file equivalence2.c line 6 column 3 function foo_original
Symex completed in: 0.006s (62 assignments)
Slicing time: 0.002s (removed 24 assignments)
Generated 5 VCC(s), 5 remaining after simplification (38 assignments)
No solver specified; defaulting to Boolector
Encoding remaining VCC(s) using bit-vector/floating-point arithmetic
Encoding to solver time: 0.008s
Solving with solver Boolector 3.2.2
Runtime decision procedure: 0.781s
BMC program time: 0.796s

VERIFICATION SUCCESSFUL

Solution found by the inductive step (k = 2)
````

If you introduce your change to the original program (i.e., `if(x <= 6 || isnan(x))`), we get the following:

````
...
Building error trace

[Counterexample]


State 1 file equivalence2.c line 36 column 3 function main thread 0
----------------------------------------------------
  N = 0 (00000000 00000000 00000000 00000000)

State 2 file equivalence2.c line 37 column 3 function main thread 0
----------------------------------------------------
  x = 2.752242e-318 (00000000 00000000 00000000 00000000 00000000 00001000 10000000 00000100)

State 4 file equivalence2.c line 38 column 3 function main thread 0
----------------------------------------------------
Violated property:
  file equivalence2.c line 38 column 3 function main
  assertion
  return_value$_foo$1 == return_value$_foo_original$2


VERIFICATION FAILED

Bug found (k = 1)
````

**Comment by lucasccordeiro** [comment]
By the way, proving the equivalence of our translation from C/C++ abstraction to the GOTO program will certainly be very useful work for ESBMC, and our community will highly recognize it.

Overall, proving program equivalence is an undecidable problem. However, you can prove many transformations within the ESBMC framework.

**Reply by hassan-truscova** [reply]
Dear @lucasccordeiro,

Thank you very much for taking out the time and answering in detail. I really appreciate it.
I now understand where my misunderstanding was.

I agree that proving program equivalence in general is an undecidable problem.
As for proving the equivalence of C/C++ to GOTO translation, well, we are bounded by time and other shiny ideas :) But i will keep it in mind and contact you via email when we start some efforts in this direction.

Thanks again.

Best wishes from Bremen,
